Shedlock - Misleading Logging of Error

As of today we see daily ERROR logs during running of deletion jobs.

Can be seen @ACC,PRD and your local instance when running over night.

Sample from ACC:
```
timestamp="2020-10-27T23:00:00.467Z", level="ERROR", hostname="tmldb00111.cc.cec.eu.int", pid="15272", thread="scheduling-1", class="o.s.s.s.TaskUtils$LoggingErrorHandler", message="Unexpected error occurred in scheduled task", exception="org.springframework.orm.ObjectOptimisticLockingFailureException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: delete from diagnosiskey where id=?; nested exception is org.hibernate.StaleStateException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: delete from diagnosiskey where id=?, 	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:340), 	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:255), 	at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:538), 	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:743), 	at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:711), 	at org.springframework.transaction.interceptor.TransactionAspectSupport.commitTransactionAfterReturning(TransactionAspectSupport.java:633), 	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:386), 	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:118), 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186), 	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:139), 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186), 	at org.springframework.data.jpa.repository.support.CrudMethodMetadataPostProcessor$CrudMethodMetadataPopulatingMethodInterceptor.invoke(CrudMethodMetadataPostProcessor.java:149), 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186), 	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:95), 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186), 	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212), 	at com.sun.proxy.$Proxy247.deleteByCreatedAtBefore(Unknown Source), 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method), 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62), 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43), 	at java.base/java.lang.reflect.Method.invoke(Method.java:566), 	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344), 	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:205), 	at com.sun.proxy.$Proxy213.deleteByCreatedAtBefore(Unknown Source), 	at eu.interop.federationgateway.service.DiagnosisKeyEntityService.deleteAllBefore(DiagnosisKeyEntityService.java:147), 	at eu.interop.federationgateway.service.DiagnosisKeyEntityService$$FastClassBySpringCGLIB$$be658b2f.invoke(<generated>), 	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218), 	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:687), 	at eu.interop.federationgateway.service.DiagnosisKeyEntityService$$EnhancerBySpringCGLIB$$482f74a9.deleteAllBefore(<generated>), 	at eu.interop.federationgateway.service.DiagnosisKeyCleanupService.cleanupDiagnosisKeys(DiagnosisKeyCleanupService.java:35), 	at eu.interop.federationgateway.service.DiagnosisKeyCleanupService$$FastClassBySpringCGLIB$$3e221d34.invoke(<generated>), 	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218), 	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:771), 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163), 	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:749), 	at org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:88), 	at org.springframework.cloud.sleuth.instrument.scheduling.TraceSchedulingAspect.traceBackgroundThread(TraceSchedulingAspect.java:76), 	at jdk.internal.reflect.GeneratedMethodAccessor491.invoke(Unknown Source), 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43), 	at java.base/java.lang.reflect.Method.invoke(Method.java:566), 	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethodWithGivenArgs(AbstractAspectJAdvice.java:644), 	at org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethod(AbstractAspectJAdvice.java:633), 	at org.springframework.aop.aspectj.AspectJAroundAdvice.invoke(AspectJAroundAdvice.java:70), 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186), 	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:749), 	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:95), 	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186), 	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:749), 	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:691), 	at eu.interop.federationgateway.service.DiagnosisKeyCleanupService$$EnhancerBySpringCGLIB$$da12d405.cleanupDiagnosisKeys(<generated>), 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method), 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62), 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43), 	at java.base/java.lang.reflect.Method.invoke(Method.java:566), 	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:84), 	at org.springframework.cloud.sleuth.instrument.async.TraceRunnable.run(TraceRunnable.java:68), 	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54), 	at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:93), 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515), 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264), 	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304), 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128), 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628), 	at java.base/java.lang.Thread.run(Thread.java:834), Caused by: org.hibernate.StaleStateException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: delete from diagnosiskey where id=?, 	at org.hibernate.jdbc.Expectations$BasicExpectation.checkBatched(Expectations.java:67), 	at org.hibernate.jdbc.Expectations$BasicExpectation.verifyOutcome(Expectations.java:54), 	at org.hibernate.engine.jdbc.batch.internal.NonBatchingBatch.addToBatch(NonBatchingBatch.java:47), 	at org.hibernate.persister.entity.AbstractEntityPersister.delete(AbstractEntityPersister.java:3572), 	at org.hibernate.persister.entity.AbstractEntityPersister.delete(AbstractEntityPersister.java:3832), 	at org.hibernate.action.internal.EntityDeleteAction.execute(EntityDeleteAction.java:124), 	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604), 	at org.hibernate.engine.spi.ActionQueue.lambda$executeActions$1(ActionQueue.java:478), 	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684), 	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:475), 	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:348), 	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:40), 	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:102), 	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1363), 	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:454), 	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3213), 	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2381), 	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:447), 	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.beforeCompletionCallback(JdbcResourceLocalTransactionCoordinatorImpl.java:183), 	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.access$300(JdbcResourceLocalTransactionCoordinatorImpl.java:40), 	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commit(JdbcResourceLocalTransactionCoordinatorImpl.java:281), 	at org.hibernate.engine.transaction.internal.TransactionImpl.commit(Transactio
class = o.s.s.s.TaskUtils$LoggingErrorHandlerhost = tmldb00111.cc.cec.eu.inthostname = tmldb00111.cc.cec.eu.intsource = /ec/local/tomcat/logs/efgs.logsourcetype = sante:efgs
```

--- a/src/main/java/eu/interop/federationgateway/service/CallbackTaskCleanupService.java
+++ b/src/main/java/eu/interop/federationgateway/service/CallbackTaskCleanupService.java
@@ -25,6 +25,7 @@ import eu.interop.federationgateway.utils.EfgsMdc;
 import java.time.ZonedDateTime;
 import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
+import net.javacrumbs.shedlock.spring.annotation.SchedulerLock;
 import org.springframework.scheduling.annotation.Scheduled;
 import org.springframework.stereotype.Component;
 
@@ -42,6 +43,8 @@ public class CallbackTaskCleanupService {
    * Removes periodically the execution locks from abandoned tasks.
    */
   @Scheduled(fixedDelay = 60000)
+  @SchedulerLock(name = "CallbackTaskCleanupService_deleteAbandonedLocks", lockAtLeastFor = "PT0S",
+    lockAtMostFor = "${efgs.callback.locklimit}")
   public void deleteAbandonedLocks() {
     log.info("Deleting task locks of abandoned tasks");
 

--- a/src/main/java/eu/interop/federationgateway/service/DiagnosisKeyCleanupService.java
+++ b/src/main/java/eu/interop/federationgateway/service/DiagnosisKeyCleanupService.java
@@ -27,6 +27,7 @@ import java.time.ZoneOffset;
 import java.time.ZonedDateTime;
 import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
+import net.javacrumbs.shedlock.spring.annotation.SchedulerLock;
 import org.springframework.scheduling.annotation.Scheduled;
 import org.springframework.stereotype.Service;
 
@@ -45,6 +46,8 @@ public class DiagnosisKeyCleanupService {
    * Cleanup task to delete all DiagnosisKeys and DiagnosisKeyBatches which are older then configured.
    */
   @Scheduled(cron = "0 0 0 * * *")
+  @SchedulerLock(name = "DiagnosisKeyCleanupService_cleanupDiagnosisKeys", lockAtLeastFor = "PT0S",
+    lockAtMostFor = "${efgs.download-settings.locklimit}")
   public void cleanupDiagnosisKeys() {
     ZonedDateTime deleteTimestamp = LocalDate.ofInstant(ZonedDateTime.now(ZoneOffset.UTC).toInstant(), ZoneOffset.UTC)
       .atStartOfDay(ZoneOffset.UTC)

--- a/src/main/resources/application.yml
+++ b/src/main/resources/application.yml
@@ -28,6 +28,7 @@ efgs:
     keyStorePass: 3fgs-p4ssw0rd
     certificateAlias: efgs_trust_anchor
   callback:
+    locklimit: 1800000
     keyStorePath: /ec/prod/app/san/efgs/efgs-cb-client.jks
     keyStorePass: 3fgs-p4ssw0rd
     keyStorePrivateKeyAlias: efgs_callback_key
@@ -48,6 +49,7 @@ efgs:
   upload-settings:
     maximum-upload-batch-size: 5000
   download-settings:
+    locklimit: 1800000
     max-age-in-days: 6
   cert-auth:
     header-fields:

--- a/src/test/resources/application.yml
+++ b/src/test/resources/application.yml
@@ -27,6 +27,7 @@ efgs:
     keyStorePass: 3fgs-p4ssw0rd
     certificateAlias: efgs_trust_anchor
   callback:
+    locklimit: 1800000
     keyStorePath: keystore/efgs-cb-client.jks
     keyStorePass: 3fgs-p4ssw0rd
     keyStorePrivateKeyAlias: efgs_callback_key
@@ -47,6 +48,7 @@ efgs:
   upload-settings:
     maximum-upload-batch-size: 5000
   download-settings:
+    locklimit: 1800000
     max-age-in-days: 14
   cert-auth:
     header-fields:

